contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,198 @@
+                        (toDistribute * (nextWeek - t)) /
+                        sinceLast;
+                }
+            }
+            t = nextWeek;
+            thisWeek = nextWeek;
+        }
+        emit CheckpointToken(block.timestamp, toDistribute);
+    }
+
+    /// @notice this function will fetch StakingRewardsV2 to see what their staked balance
+    /// was at the start of the epoch then calculate proportional fees and transfer to user
+    function claimEpoch(address to, uint epochNumber) public returns (uint256) {
+        /// @dev if more than 24 hours from last checkpoint OR if it is the first
+        /// claim of the week. second condition is so that the end of a week always
+        /// gets updated before its claimed.
+        if (
+            (block.timestamp - lastCheckpoint > 86400) ||
+            ((block.timestamp - startOfWeek(lastCheckpoint)) > 604800)
+        ) {
+            checkpointToken();
+        }
+        /// @dev if the end of the epoch is > current time, revert
+        if ((epochNumber * 1 weeks) + 1 weeks + startTime > block.timestamp) {
+            revert CannotClaimYet();
+        }
+        if (claimedEpochs[to][epochNumber] == true) {
+            revert CannotClaimTwice();
+        }
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(
+            (epochNumber * 1 weeks) + startTime
+        );
+        if (totalStaked == 0) {
+            revert NothingStakedThatEpoch();
+        }
+
+        uint256 proportionalFees = calculateEpochFees(to, epochNumber);
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+        claimedEpochs[to][epochNumber] = true;
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
+
+        return proportionalFees;
+    }
+
+    /// @notice claim many epochs at once
+    function claimMany(address to, uint[] memory epochs) public {
+        for (uint i = 0; i < epochs.length; i++) {
+            uint epochNumber = epochs[i];
+            claimEpoch(to, epochNumber);
+        }
+    }
+
+    /// @notice view function for calculating fees for an epoch
+    function calculateEpochFees(
+        address to,
+        uint epochNumber
+    ) public view returns (uint256) {
+        uint thisWeek = (epochNumber * 1 weeks) + startTime;
+        uint256 userStaked = stakingRewardsV2.balanceAtTime(to, thisWeek);
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(thisWeek);
+        if (totalStaked == 0) {
+            return totalStaked;
+        }
+        uint256 proportionalFees = ((tokensPerEpoch[thisWeek] * userStaked) /
+            totalStaked);
+
+        return proportionalFees;
+    }
+
+    /// @notice function for calculating the start of a week with an offset
+    function startOfWeek(uint timestamp) public view returns (uint) {
```

Comment:
Should be internal if never planned to be called publicly

ie. utility functions don't need to be public